### ISSUE TITLE
Syntax errors should be reported quickly

### ISSUE THREAD

**Issue by juliemr** [dart-lang/test#290]
When I have a syntax error in a test, e.g.

```
test('do foo', () {
}       // oops forgot to close the parens
```

instead of failing quickly, the test waits for a minute and then gives

```
Failed to load "test/your_test.dart": Timed out waiting for the test suite to connect on Dartium.
```

It would be great if the error could be reported immediately and note that it's a syntax error.

**Comment by nex3** [comment]
This is a surprisingly tough problem. Dartium doesn't expose syntax errors in any way that's accessible to the host page, so it's very hard to get access to them and associate them with the page that's loading. It's possible that talking to the VM service will let this happen, but that's a ways away.

**Comment by skybrian** [comment]
This isn't just a dartium issue; it also happened to me a couple of times with a regular vm test. (In my case, the problem was a change marker I hadn't removed during a merge.)

A possible solution might be for "pub run test" to do syntax checking in parallel with the tests? If the test hangs, there's plenty of time to finish the syntax check. Otherwise it can be aborted.

**Comment by nex3** [comment]
> This isn't just a dartium issue; it also happened to me a couple of times with a regular vm test. (In my case, the problem was a change marker I hadn't removed during a merge.)

At a guess, this is a side effect of dart-lang/sdk#23684.

> A possible solution might be for "pub run test" to do syntax checking in parallel with the tests? If the test hangs, there's plenty of time to finish the syntax check. Otherwise it can be aborted.

That's possible, but it would be quite complex, and it would still likely slow down the loading. The test runner is already heavily parallel, so adding more parallelism would come with a speed cost.

**Comment by skybrian** [comment]
A possible way to fix that might be wait until the tests have been running for a few seconds, and skip the test files that have already finished. Then you only pay the cost when there are tests that are probably hung and the test suite won't finish quickly anyway.

**Comment by nex3** [comment]
That would also be possible, but complicated. I'd rather fix this at the source by solving dart-lang/sdk#23684 and talking to the VM service to catch Dartium errors.

**Comment by yjbanov** [comment]
Can the test runner detect that an isolate was killed without reporting back anything reasonable? If so, it could print something descriptive, such as `Test isolate crashed unexpectedly, perhaps due to a syntax error in your code (see https://github.com/dart-lang/test/issues/290)`, or something along those lines.

I lost quite a bit of time today because of this. Would love to see at least a temporary fix.

The workaround is to run the test using plain `dart` command.

**Comment by nex3** [comment]
@yjbanov For browser tests, we have absolutely no way to distinguish a test failing due to a syntax error from one that's just taking a long time to connect. We can't even see that it crashed. The best we can do is time out, and we already have timeouts in place that are about as narrow as they can get without causing flakiness.

**Comment by yjbanov** [comment]
My case is actually a plain VM test. As for browser, perhaps you could use something similar to what [webdriver](https://github.com/google/dart-sync-webdriver/blob/c53682fa50bbc5baf304b9c096b396c5e1632612/lib/src/logs.dart#L26) does to extract log entries along with severity levels, and simply print the errors to the console so that it is more evident that something beyond a timeout is going on.

**Comment by skybrian** [comment]
One thing that might help would be for the test runner to report back even more often?

The first thing that the isolate should do is report back that main() was successfully entered. This can have a relatively short timeout, and if it fails, we can report that there was some kind of problem in the load phase, perhaps a syntax error.

Looks like the relevant code is here? Perhaps add another response type for "started".
https://github.com/dart-lang/test/blob/master/lib/src/runner/loader.dart#L204

(I don't know if the hang actually happens in that code, but a similar principle applies for other operations that might cause the test runner to lose contact with the caller: give the flaky operation its own status message in the communication protocol. It doesn't need to be logged unless the pingback isn't received in a reasonable amount of time.)

**Comment by nex3** [comment]
I'm going to split this into two separate issues, since the underlying problems and potential solutions are so different. I'm retargeting this one to be specifically about syntax errors on Dartium/content shell, since that was Julie's original issue, and I've filed #312 to track the handling of runtime syntax errors on the VM. I'll continue the conversation there.

**Comment by nex3** [comment]
> As for browser, perhaps you could use something similar to what webdriver does to extract log entries along with severity levels, and simply print the errors to the console so that it is more evident that something beyond a timeout is going on.

As far as I can tell, Dartium doesn't print Dart syntax errors to the console either.

**Comment by yjbanov** [comment]
I certainly _see_ syntax errors in the dev tools console, although I admit that don't know the mechanism by which those errors get there.

**Comment by nex3** [comment]
Ah, I thought you meant via standard out or standard error. As far as I know, there's no way to capture console log either, though.

**Comment by yjbanov** [comment]
Looking at https://github.com/google/webdriver.dart/blob/master/lib/src/logs.dart#L21 it seems you can get logs simply by polling.

@DrMarcII: do you know if it's possible to extract browser logs without resorting to big infrastructures (e.g. not having to run webdriver)?

**Comment by DrMarcII** [comment]
I believe if you launch chrome with the flag --enable-logging (possibly with increased verbosity), it will write logs (including console output) to the chrome_debug.log in the user data directory (https://www.chromium.org/for-testers/enable-logging).

**Comment by kevmoo** [comment]
FYI: SDK Fixed https://github.com/dart-lang/sdk/issues/23684
